validator/src/journal/chain_commit_state.rs

```diff
@@ -104,6 +104,122 @@ impl<BS: IndexedBlockStore> ChainCommitState<BS> {
         let last_block = blocks.last().cloned();
         (batch_ids, txn_ids, last_block)
     }
+
+    fn block_in_chain(&self, block: &Block) -> bool {
+        if let Some(ref common) = self.ancestor {
+            return block.block_num <= common.block_num;
+        }
+        false
+    }
+
+    pub fn validate_no_duplicate_batches(
+        &self,
+        batch_ids: &[&str],
+    ) -> Result<(), ChainCommitStateError> {
+        if let Some(batch_id) = check_no_duplicates(batch_ids) {
+            return Err(ChainCommitStateError::DuplicateBatch(batch_id));
+        }
+        for id in batch_ids {
+            if self.uncommitted_batch_ids.contains(&(*id).into()) {
```

Comment:
You should have to do this: you are turning a reference into a `String` and then a `&str`

Reply:
Yeah, I agree it looks a little funny. if I just dereference the `&&str` I get this error
```error[E0308]: mismatched types
   --> src/journal/chain_commit_state.rs:156:50
    |
156 |             if self.uncommitted_txn_ids.contains(*id) {
    |                                                  ^^^ expected struct `std::string::String`, found str
    |
    = note: expected type `&std::string::String`
               found type `&str`
```